plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java

```diff
@@ -304,6 +316,94 @@ private static String quotedName(String name)
+    {
+        return table.specs().values().stream()
+                .allMatch(spec -> canEnforceConstraintWithinPartitioningSpec(typeOperators, spec, columnHandle, domain));
+    }
+
+    private static boolean canEnforceConstraintWithinPartitioningSpec(TypeOperators typeOperators, PartitionSpec spec, IcebergColumnHandle column, Domain domain)
+    {
+        for (PartitionField field : spec.fields()) {
+            if (field.transform().toString().equals("void")) {
+                // Useless for filtering.
+                continue;
+            }
+            if (field.sourceId() != column.getId()) {
+                // Not related to each other.
+                continue;
+            }
+            if (field.transform().isIdentity()) {
+                // A predicate on an identity partitioning column can always be enforced.
+                return true;
+            }
+
+            // Partitioning transform must return NULL for NULL input. Here we assume it
+            // never returns NULL for non-NULL input (unless it's the void transform),
+            // so NULL values and non-NULL values are always segregated.
+            ValueSet valueSet = domain.getValues();
+
+            boolean canEnforce = valueSet.getValuesProcessor().transform(
+                    ranges -> {
+                        ColumnTransform transform = PartitionTransforms.getColumnTransform(field, column.getType());
+                        MethodHandle targetTypeEqualOperator = typeOperators.getEqualOperator(
+                                transform.getType(), InvocationConvention.simpleConvention(FAIL_ON_NULL, NEVER_NULL, NEVER_NULL));
+                        for (Range range : ranges.getOrderedRanges()) {
+                            if (!canEnforceRangeWithPartitioningField(transform, range, targetTypeEqualOperator)) {
+                                return false;
+                            }
+                        }
+                        return true;
+                    },
+                    discreteValues -> false,
+                    allOrNone -> true);
+            if (canEnforce) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static boolean canEnforceRangeWithPartitioningField(ColumnTransform transform, Range range, MethodHandle targetTypeEqualOperator)
+    {
+        io.trino.spi.type.Type type = range.getType();
+        if (!type.isOrderable()) {
+            return false;
+        }
+        if (!range.isLowUnbounded()) {
+            Object boundedValue = range.getLowBoundedValue();
+            Optional<Object> adjacentValue = range.isLowInclusive() ? getPreviousValue(type, boundedValue) : getNextValue(type, boundedValue);
+            if (adjacentValue.isEmpty() || yieldSamePartitioningValue(type, transform, boundedValue, adjacentValue.get(), targetTypeEqualOperator)) {
```

Comment:
The logic is supposed to be suitable for monotonic partitioning transforms, but it is not suitable for eg bucketing.